diff -up dhcp-4.1.0/client/dhclient.c.semicolons dhcp-4.1.0/client/dhclient.c
--- dhcp-4.1.0/client/dhclient.c.semicolons	2008-06-11 10:17:10.000000000 -1000
+++ dhcp-4.1.0/client/dhclient.c	2009-08-04 17:46:38.000000000 -1000
@@ -2780,7 +2780,7 @@ write_client6_lease(struct client_state 
 			if (addr->options != NULL)
 				write_options(client, addr->options, "      ");
 
-			stat = fprintf(leaseFile, "    }\n");
+			stat = fprintf(leaseFile, "    };\n");
 			if (stat <= 0)
 				return ISC_R_IOERROR;
 		}
@@ -2788,7 +2788,7 @@ write_client6_lease(struct client_state 
 		if (ia->options != NULL)
 			write_options(client, ia->options, "    ");
 
-		stat = fprintf(leaseFile, "  }\n");
+		stat = fprintf(leaseFile, "  };\n");
 		if (stat <= 0)
 			return ISC_R_IOERROR;
 	}
@@ -2802,7 +2802,7 @@ write_client6_lease(struct client_state 
 	if (lease->options != NULL)
 		write_options(client, lease->options, "  ");
 
-	stat = fprintf(leaseFile, "}\n");
+	stat = fprintf(leaseFile, "};\n");
 	if (stat <= 0)
 		return ISC_R_IOERROR;
 
@@ -2926,20 +2926,20 @@ int write_client_lease (client, lease, r
 
 	tval = print_time(lease->renewal);
 	if (tval == NULL ||
-	    fprintf(leaseFile, "  renew %s\n", tval) < 0)
+	    fprintf(leaseFile, "  renew %s;\n", tval) < 0)
 		errors++;
 
 	tval = print_time(lease->rebind);
 	if (tval == NULL ||
-	    fprintf(leaseFile, "  rebind %s\n", tval) < 0)
+	    fprintf(leaseFile, "  rebind %s;\n", tval) < 0)
 		errors++;
 
 	tval = print_time(lease->expiry);
 	if (tval == NULL ||
-	    fprintf(leaseFile, "  expire %s\n", tval) < 0)
+	    fprintf(leaseFile, "  expire %s;\n", tval) < 0)
 		errors++;
 
-	if (fprintf(leaseFile, "}\n") < 0)
+	if (fprintf(leaseFile, "};\n") < 0)
 		errors++;
 
 	if (fflush(leaseFile) != 0)
