From ed7aa33898bfcf5a37f5b730897611a18f6a929b Mon Sep 17 00:00:00 2001
From: Jiri Popelka <jpopelka@redhat.com>
Date: Tue, 7 Jan 2014 12:13:46 +0100
Subject: [PATCH] missing comma in ldap.c

---
 server/ldap.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/server/ldap.c b/server/ldap.c
index efbc874..2994106 100644
--- a/server/ldap.c
+++ b/server/ldap.c
@@ -628,7 +628,7 @@ ldap_start (void)
       options = NULL;
       option_state_allocate (&options, MDL);
 
-      execute_statements_in_scope (NULL, NULL, NULL, NULL, NULL
+      execute_statements_in_scope (NULL, NULL, NULL, NULL, NULL,
 				   options, &global_scope, root_group,
 				   NULL, NULL);
 
-- 
1.8.4.2

