diff -up dhcp-4.0.0/common/discover.c.fp dhcp-4.0.0/common/discover.c
--- dhcp-4.0.0/common/discover.c.fp	2008-01-15 10:25:45.000000000 -1000
+++ dhcp-4.0.0/common/discover.c	2008-01-15 10:27:08.000000000 -1000
@@ -461,6 +461,12 @@ next_iface4(struct iface_info *info, int
 	char *name;
 	struct ifreq tmp;
 
+	if (!ifaces->fp) {
+		log_error("Error reading interface information");
+		*err = 1;
+		return 0;
+	}
+
 	/*
 	 * Loop exits when we find an interface that has an address, or 
 	 * when we run out of interfaces.
@@ -587,6 +593,12 @@ next_iface6(struct iface_info *info, int
 	struct sockaddr_in6 addr;
 	struct ifreq tmp;
 
+	if (!ifaces->fp6) {
+		log_error("Error reading interface information");
+		*err = 1;
+		return 0;
+	}
+
 	do {
 		/*
 		 *  Read the next line in the file.
