--- dhcp-3.0.5/client/dhclient.c.usage	2007-04-17 10:43:00.000000000 -0400
+++ dhcp-3.0.5/client/dhclient.c	2007-04-17 10:43:30.000000000 -0400
@@ -819,20 +819,29 @@
 
 static void usage ()
 {
-	log_info ("%s %s", message, DHCP_VERSION);
-	log_info (copyright);
-	log_info (arr);
-	log_info (url);
+	printf ("%s %s\n", message, DHCP_VERSION);
+	printf (copyright);
+	printf ("\n");
+	printf (arr);
+	printf ("\n");
+	printf (url);
+	printf ("\n");
 
 #ifdef EXTENDED_NEW_OPTION_INFO
-	log_error ("Usage: dhclient [-1dqr] [-nwx] [-p <port>] %s",
+	printf ("Usage: dhclient [-1dqr] [-nwx] [-p <port>] %s",
 #else
-	log_error ("Usage: dhclient [-1dqr] [-nw] [-p <port>] %s",
+	printf ("Usage: dhclient [-1dqr] [-nw] [-p <port>] %s",
 #endif
-		   "[-s server]");
-	log_error ("                [-cf config-file] [-lf lease-file]%s",
-		   "[-pf pid-file] [-e VAR=val]");
-	log_fatal ("                [-sf script-file] [interface]");
+	   "[-s server]");
+	printf ("                [-cf config-file] [-lf lease-file]%s",
+	   "[-pf pid-file] [-e VAR=val]");
+	printf ("                [ -I <dhcp-client-identifier> ] [-B]\n");
+	printf ("                [ -H <host-name> | -F <fqdn.fqdn> ] [ -T <timeout> ]\n");
+	printf ("                [ -V <vendor-class-identifier> ]\n");
+	printf ("                [ -R <request option list> ]\n");
+	printf ("                [-sf script-file] [interface]\n");
+
+	exit (EXIT_FAILURE);
 }
 
 isc_result_t find_class (struct class **c,
