diff -up dhcp-3.0.6/client/dhclient.c.enoi dhcp-3.0.6/client/dhclient.c
--- dhcp-3.0.6/client/dhclient.c.enoi	2007-08-15 16:14:32.000000000 -0400
+++ dhcp-3.0.6/client/dhclient.c	2007-08-15 16:17:17.000000000 -0400
@@ -259,6 +259,9 @@ int main (argc, argv, envp)
 		} else if (!strcmp (argv [i], "--version")) {
 			log_info ("isc-dhclient-%s", DHCP_VERSION);
 			exit (0);
+		} else if (!strcmp (argv [i], "-x")) {
+			/* ignored for compatibility */
+			continue;
 		} else if (!strcmp (argv [i], "-I")) {
 			if ((++i == argc) || (argv[i] == NULL) || (*(argv[i])=='\0')) {
 				usage ();
