diff -up dhcp-4.2.3-P2/client/dhclient.c.send_release dhcp-4.2.3-P2/client/dhclient.c
--- dhcp-4.2.3-P2/client/dhclient.c.send_release	2012-03-21 19:16:16.000000000 +0100
+++ dhcp-4.2.3-P2/client/dhclient.c	2012-03-21 19:21:39.920599903 +0100
@@ -2749,8 +2749,8 @@ void send_release (cpp)
 	      inet_ntoa (destination.sin_addr),
 	      ntohs (destination.sin_port), client -> xid);
 
-	if (fallback_interface)
-		result = send_packet (fallback_interface,
+	if (client -> interface)
+		result = send_packet (client -> interface,
 				      (struct packet *)0,
 				      &client -> packet,
 				      client -> packet_length,
@@ -2758,7 +2758,7 @@ void send_release (cpp)
 				      (struct hardware *)0);
 	else
 		/* Send out a packet. */
-		result = send_packet (client -> interface, (struct packet *)0,
+		result = send_packet (fallback_interface, (struct packet *)0,
 				      &client -> packet,
 				      client -> packet_length,
 				      from, &destination,
