diff -up dhcp-4.1.1/client/dhclient.c.initialization-delay dhcp-4.1.1/client/dhclient.c
--- dhcp-4.1.1/client/dhclient.c.initialization-delay	2010-04-29 17:02:09.000000000 +0200
+++ dhcp-4.1.1/client/dhclient.c	2010-04-29 17:03:27.000000000 +0200
@@ -930,7 +930,7 @@ main(int argc, char **argv) {
 					/* Set up a timeout to start the
 					 * initialization process.
 					 */
-					tv.tv_sec = cur_time + random() % 5;
+					tv.tv_sec = cur_time + random() % 1;
 					tv.tv_usec = 0;
 					add_timeout(&tv, state_reboot,
 						    client, 0, 0);
@@ -3913,7 +3913,7 @@ isc_result_t dhclient_interface_startup_
 			client -> state = S_INIT;
 			/* Set up a timeout to start the initialization
 			   process. */
-			tv . tv_sec = cur_time + random () % 5;
+			tv . tv_sec = cur_time + random () % 1;
 			tv . tv_usec = 0;
 			add_timeout (&tv, state_reboot, client, 0, 0);
 		}
