diff -up dhcp-4.0.0//server/confpars.c.failover dhcp-4.0.0//server/confpars.c
--- dhcp-4.0.0//server/confpars.c.failover	2008-08-01 10:07:46.000000000 -1000
+++ dhcp-4.0.0//server/confpars.c	2008-08-01 10:14:03.000000000 -1000
@@ -1065,10 +1065,17 @@ void parse_failover_peer (cfile, group, 
 		parse_warn (cfile, "peer address may not be omitted");
 
 	/* XXX - when/if we get a port number assigned, just set as default */
+	/* See Red Hat Bugzilla 167292:
+	 * we do now:  dhcp-failover 647/tcp
+	 *             dhcp-failover 647/udp
+	 *             dhcp-failover 847/tcp
+	 *             dhcp-failover 847/udp
+	 * IANA registration by Bernard Volz <volz@cisco.com>
+	 */
 	if (!peer -> me.port)
-		parse_warn (cfile, "local port may not be omitted");
+		peer -> me.port = 647;
 	if (!peer -> partner.port)
-		parse_warn (cfile, "peer port may not be omitted");
+		peer -> partner.port = 847;
 
 	if (peer -> i_am == primary) {
 	    if (!peer -> hba) {
