diff -up dhcp-4.0.0/includes/dhcpd.h.path dhcp-4.0.0/includes/dhcpd.h
--- dhcp-4.0.0/includes/dhcpd.h.path	2008-08-23 18:25:54.000000000 -1000
+++ dhcp-4.0.0/includes/dhcpd.h	2008-08-23 18:26:21.000000000 -1000
@@ -1302,15 +1302,15 @@ typedef unsigned char option_mask [16];
 #else /* !DEBUG */
 
 #ifndef _PATH_DHCPD_CONF
-#define _PATH_DHCPD_CONF	"/etc/dhcpd.conf"
+#define _PATH_DHCPD_CONF	"/etc/dhcp/dhcpd.conf"
 #endif /* DEBUG */
 
 #ifndef _PATH_DHCPD_DB
-#define _PATH_DHCPD_DB		LOCALSTATEDIR"/db/dhcpd.leases"
+#define _PATH_DHCPD_DB		LOCALSTATEDIR"/dhcpd/dhcpd.leases"
 #endif
 
 #ifndef _PATH_DHCPD6_DB
-#define _PATH_DHCPD6_DB		LOCALSTATEDIR"/db/dhcpd6.leases"
+#define _PATH_DHCPD6_DB		LOCALSTATEDIR"/dhcpd/dhcpd6.leases"
 #endif
 
 #ifndef _PATH_DHCPD_PID
@@ -1324,7 +1324,7 @@ typedef unsigned char option_mask [16];
 #endif /* DEBUG */
 
 #ifndef _PATH_DHCLIENT_CONF
-#define _PATH_DHCLIENT_CONF	"/etc/dhclient.conf"
+#define _PATH_DHCLIENT_CONF	"/etc/dhcp/dhclient.conf"
 #endif
 
 #ifndef _PATH_DHCLIENT_SCRIPT
@@ -1340,11 +1340,11 @@ typedef unsigned char option_mask [16];
 #endif
 
 #ifndef _PATH_DHCLIENT_DB
-#define _PATH_DHCLIENT_DB	LOCALSTATEDIR"/db/dhclient.leases"
+#define _PATH_DHCLIENT_DB	LOCALSTATEDIR"/dhclient/dhclient.leases"
 #endif
 
 #ifndef _PATH_DHCLIENT6_DB
-#define _PATH_DHCLIENT6_DB	LOCALSTATEDIR"/db/dhclient6.leases"
+#define _PATH_DHCLIENT6_DB	LOCALSTATEDIR"/dhclient/dhclient6.leases"
 #endif
 
 #ifndef _PATH_RESOLV_CONF
