diff -up dhcp-4.1.1/configure.ac.DSO dhcp-4.1.1/configure.ac
--- dhcp-4.1.1/configure.ac.DSO	2010-02-16 11:52:30.000000000 +0100
+++ dhcp-4.1.1/configure.ac	2010-02-16 11:52:55.000000000 +0100
@@ -548,7 +548,7 @@ AC_ARG_WITH(ldap,
     [ldap=no])
 if test x$ldap = xyes ; then
     AM_CHECK_LDAP
-    ldap_libs="-lldap"
+    ldap_libs="-lldap -llber"
     AC_SUBST(LDAP_CFLAGS, [-DLDAP_CONFIGURATION])
     AC_SUBST(LDAP_LIBS, [$ldap_libs])
 fi
