diff -up dhcp-4.1.0p1/client/dhclient.c.logpid dhcp-4.1.0p1/client/dhclient.c
--- dhcp-4.1.0p1/client/dhclient.c.logpid	2009-12-14 11:39:47.000000000 +0100
+++ dhcp-4.1.0p1/client/dhclient.c	2009-12-14 11:41:17.000000000 +0100
@@ -148,7 +148,7 @@ main(int argc, char **argv) {
 	else if (fd != -1)
 		close(fd);
 
-	openlog("dhclient", LOG_NDELAY, LOG_DAEMON);
+	openlog("dhclient", LOG_NDELAY | LOG_PID, LOG_DAEMON);
 
 #if !(defined(DEBUG) || defined(__CYGWIN32__))
 	setlogmask(LOG_UPTO(LOG_INFO));
