diff -up dhcp-4.1.0/common/tables.c.garbage dhcp-4.1.0/common/tables.c
--- dhcp-4.1.0/common/tables.c.garbage	2008-01-23 16:43:04.000000000 -1000
+++ dhcp-4.1.0/common/tables.c	2009-01-06 12:09:41.000000000 -1000
@@ -207,7 +207,7 @@ static struct option dhcp_options[] = {
 	{ "netinfo-server-tag", "t",		&dhcp_universe, 113, 1 },
 	{ "default-url", "t",			&dhcp_universe, 114, 1 },
 	{ "subnet-selection", "I",		&dhcp_universe, 118, 1 },
-	{ "domain-search", "Dc",		&dhcp_universe, 119, 1 },
+	{ "domain-search", "D",		&dhcp_universe, 119, 1 },
 	{ "vivco", "Evendor-class.",		&dhcp_universe, 124, 1 },
 	{ "vivso", "Evendor.",			&dhcp_universe, 125, 1 },
 #if 0
