diff -up dhcp-4.0.0//client/dhclient.c.fast_timeout dhcp-4.0.0//client/dhclient.c
--- dhcp-4.0.0//client/dhclient.c.fast_timeout	2008-08-01 10:10:10.000000000 -1000
+++ dhcp-4.0.0//client/dhclient.c	2008-08-01 10:13:37.000000000 -1000
@@ -3613,10 +3613,7 @@ isc_result_t dhclient_interface_startup_
 		ip -> flags |= INTERFACE_RUNNING;
 		for (client = ip -> client; client; client = client -> next) {
 			client -> state = S_INIT;
-			/* Set up a timeout to start the initialization
-			   process. */
-			add_timeout (cur_time + random () % 5,
-				     state_reboot, client, 0, 0);
+			add_timeout (cur_time, state_reboot, client, 0, 0);
 		}
 	}
 	return ISC_R_SUCCESS;
